### PR TITLE
Add custom install and postinstall hooks

### PR DESCRIPTION
**Pull request by leeweisberger**
Add custom hooks accessible as:
```
airplane: {
  "install":
  "postinstall":
}
```

Install allows you to specify a custom installation script. e.g. if you want to install dev dependencies or do something else custom.

PostInstall allows you to run commands after install and before the build. e.g. generating a prisma client after the prisma cli has been installed.

Eventually I think we might want to (additionally) surface this in the deploy configuration that @rose-huang is working on.

### CODE REVIEW

File: pkg/build/node.go [hunk 1]
```diff
@@ -130,6 +141,7 @@ func node(root string, options KindOptions) (string, error) {
 	// in which case we could introduce an extra step for performing build commands.
 	return applyTemplate(heredoc.Doc(`
 		FROM {{.Base}}
+		ENV NODE_ENV=production
```
Comment (leeweisberger) [inline pkg/build/node.go:144]: Might as well add this. Used by different node frameworks to optimize production builds

File: pkg/build/node.go [hunk 2]
```diff
@@ -157,22 +169,21 @@ func node(root string, options KindOptions) (string, error) {
...
+
+		{{if .PostInstallCommand}}
+		RUN {{.PostInstallCommand}}
+		{{end}}
 		
 		RUN {{.InlineShim}} > /airplane/.airplane/shim.js && \
 			esbuild /airplane/.airplane/shim.js \
 				--bundle \
 				--external:airplane \
+				--minify \
```
Comment (leeweisberger) [inline pkg/build/node.go:186]: Another unrelated change. Slightly smaller builds.

File: pkg/runtime/runtime.go [hunk 3]
```diff
@@ -32,11 +32,6 @@ var (
 	ErrNotImplemented = errors.New("runtime: not implemented")
 )
 
-// Settings represent Airplane specific settings.
-type Settings struct {
```
Comment (leeweisberger) [inline pkg/runtime/runtime.go:36]: circular dep issues. Moved to /build

File: pkg/build/node.go [hunk 4]
```diff
@@ -157,22 +169,21 @@ func node(root string, options KindOptions) (string, error) {
...
 		{{end}}
 
-		{{if .IsYarn}}
-		RUN yarn --non-interactive --production --frozen-lockfile
-		{{else if .HasPackageLock}}
-		RUN npm ci --production
-		{{else}}
-		RUN npm i --production
-		{{end}}
+		RUN {{.InstallCommand}}
```
Comment (colinking) [inline pkg/build/node.go:172]: What's the use-case for the custom install command?

File: pkg/build/node.go [hunk 5]
```diff
@@ -115,6 +114,18 @@ func node(root string, options KindOptions) (string, error) {
 	}
 	cfg.InlineShim = inlineString(shim)
 
+	installCommand := "npm install --production"
+	if pkg.Settings.InstallCommand != "" {
+		installCommand = pkg.Settings.InstallCommand
```
Comment (colinking) [inline pkg/build/node.go:119]: Could you `s/\n/\\n`? Otherwise folks will hit issues if they try to add new lines as they'll get interpreted as Dockerfile commands.

File: pkg/build/node.go (hunk 2 above)
Comment (colinking) [reply pkg/build/node.go:186]: This will effect error stack traces, since they'll reference minified function names

File: pkg/build/node.go (hunk 4 above)
Comment (colinking) [reply pkg/build/node.go:172]: Is the idea that Prisma users would override this with:

```
cp prisma . && yarn install --non-interactive --frozen-lockfile
```

Basically (1) copy over whatever files are needed for `prisma generate` then (2) override our install command to install dev deps since `@prisma/client` will be a dev dep?

My understanding is they would not need a `postinstall` command since the prisma client runs `prisma generate` from a standard `package.json` `postinstall` hook.

Given all this, could you add a test case that uses Prisma? I agree it'll be a common Airplane use-case.

File: pkg/examples/typescript/yarn/package.json [hunk 6]
```diff
@@ -2,4 +2,5 @@
   "dependencies": {
     "airplane": "^0.0.5"
   }
+
```
Comment (colinking) [inline pkg/examples/typescript/yarn/package.json:5]: typo?

**Review by colinking — APPROVED** [review]

File: pkg/build/node.go (hunk 2 above)
Comment (leeweisberger) [reply pkg/build/node.go:186]: ah good call. Let me revert this then. Not worth

File: pkg/build/node.go (hunk 4 above)
Comment (leeweisberger) [reply pkg/build/node.go:172]: That was what I was thinking - they can override this to install dev deps.

They would also need a postinstall hook b/c when we run install, we don't have the rest of their code in the container yet (for caching purposes) so they'd need to manually run prisma generate.

I was also thinking we might want to auto-detect prisma and just do this correctly by default, but I think I'd prefer to add the ability to customize before we start auto-detecting

Comment (leeweisberger) [reply pkg/build/node.go:172]: Added a test here that shows the workflow: https://github.com/airplanedev/lib/pull/27/commits/adc5f6a40a1f83548935da4b019f6164c82d8a74

I will add this stuff to our documentation.
